### ISSUE TITLE
route change interceptors

### ISSUE THREAD

**Issue by kyeotic** [kyeotic/raviger#23]
We need a way to confirm route changes before they happen. We need _interceptors_. I believe this may be the last feature before going to 1.0.

I am currently working on an API, but my home development time has been severely restricted the last two weeks due to a neck injury. Any ideas or sub-feature requests for interceptors are welcome here.

## Confirmation

In the confirmation case it is expected that the the intercept-handler will gather user input and decide whether to cancel or allow the navigation to occur. 

This could be achieved either through immediate cancellation that is resumed if the intercept-handler indicates, or through synchronous/browser-managed confirmation. However, according to the MDN

> To combat unwanted pop-ups, some browsers don't display prompts created in beforeunload event handlers unless the page has been interacted with. Moreover, some don't display them at all.

I believe using the browser built-ins is the right way to go, even if some browsers don't support it (that is an intentional decision we would be fighting).

## Inspection

In the inspection case it is expected a component may want to conditionally alter navigation changes or block navigation. Blocking navigation without a prompt may not be possible in all situations, but should be possible if navigation was triggered from within raviger.

### Simple

Provide a confirmation prompt that will be presented to the user if they try to navigate. Only active if `predicate` is true.

```typescript
function useInterceptor(prompt: string, predicate = true: boolean) : void
```

### Functional Sync

Provide a function that will receive the current route and a function that will cancel the transition. Must handle syncronously.

```typescript
function useInterceptor(interceptFn: (route: string, cancelFn: () => void) => void) : void
```

### Functional Async

Provide a function that will receive the current route and returns the intended route as a promise. can return the same route to allow, or `null` to cancel.

```typescript
function useInterceptor(interceptFn: (route: string) => Promise<string | null>) : void
```

**Comment by kyeotic** [comment]
Initial testing confirms [most browsers no longer allow you to control the prompt](https://stackoverflow.com/questions/38879742/is-it-possible-to-display-a-custom-message-in-the-beforeunload-popup) you present to a user. It seems like the warning message should be optional, since its rarely even used.

**Comment by kyeotic** [comment]
published under `0.5.9`
